pkg/block/indexheader/lazy_binary_reader.go

```diff
@@ -0,0 +1,228 @@
+	logger log.Logger,
+	bkt objstore.BucketReader,
+	dir string,
+	id ulid.ULID,
+	postingOffsetsInMemSampling int,
+	metrics *LazyBinaryReaderMetrics,
+) (*LazyBinaryReader, error) {
+	filepath := filepath.Join(dir, id.String(), block.IndexHeaderFilename)
+
+	// If the index-header doesn't exist we should download it.
+	if _, err := os.Stat(filepath); err != nil {
+		if !os.IsNotExist(err) {
+			return nil, errors.Wrap(err, "read index header")
+		}
+
+		level.Debug(logger).Log("msg", "the index-header doesn't exist on disk; recreating", "path", filepath)
+
+		start := time.Now()
+		if err := WriteBinary(ctx, bkt, id, filepath); err != nil {
+			return nil, errors.Wrap(err, "write index header")
+		}
+
+		level.Debug(logger).Log("msg", "built index-header file", "path", filepath, "elapsed", time.Since(start))
+	}
+
+	return &LazyBinaryReader{
+		ctx:                         ctx,
+		logger:                      logger,
+		bkt:                         bkt,
+		dir:                         dir,
+		filepath:                    filepath,
+		id:                          id,
+		postingOffsetsInMemSampling: postingOffsetsInMemSampling,
+		metrics:                     metrics,
+	}, nil
+}
+
+func (r *LazyBinaryReader) Close() error {
+	r.readerMx.Lock()
+	defer r.readerMx.Unlock()
+
+	if r.reader == nil {
+		return nil
+	}
+
+	r.metrics.unloadCount.Inc()
+	if err := r.reader.Close(); err != nil {
+		r.metrics.unloadFailedCount.Inc()
+		return err
+	}
+
+	r.reader = nil
+
+	return nil
+}
+
+func (r *LazyBinaryReader) IndexVersion() (int, error) {
+	r.readerMx.RLock()
+	defer r.readerMx.RUnlock()
+
+	if err := r.open(); err != nil {
+		return 0, err
+	}
+
+	return r.reader.IndexVersion()
+}
+
+func (r *LazyBinaryReader) PostingsOffset(name string, value string) (index.Range, error) {
+	r.readerMx.RLock()
+	defer r.readerMx.RUnlock()
+
+	if err := r.open(); err != nil {
+		return index.Range{}, err
+	}
+
+	return r.reader.PostingsOffset(name, value)
+}
+
+func (r *LazyBinaryReader) LookupSymbol(o uint32) (string, error) {
+	r.readerMx.RLock()
+	defer r.readerMx.RUnlock()
+
+	if err := r.open(); err != nil {
+		return "", err
+	}
+
+	return r.reader.LookupSymbol(o)
+}
+
+func (r *LazyBinaryReader) LabelValues(name string) ([]string, error) {
+	r.readerMx.RLock()
+	defer r.readerMx.RUnlock()
+
+	if err := r.open(); err != nil {
+		return nil, err
+	}
+
+	return r.reader.LabelValues(name)
+}
+
+func (r *LazyBinaryReader) LabelNames() ([]string, error) {
+	r.readerMx.RLock()
+	defer r.readerMx.RUnlock()
+
+	if err := r.open(); err != nil {
+		return nil, err
+	}
+
+	return r.reader.LabelNames()
+}
+
+// open ensure the underlying binary index-header reader has been successfully opened. Returns
+// an error on failure. This function MUST be called with the read lock already acquired.
+func (r *LazyBinaryReader) open() error {
+	// Nothing to do if we already tried opening it.
+	if r.reader != nil {
+		return nil
+	} else if r.readerErr != nil {
```

Comment:
no need for `else` (:

Reply:
https://thanos.io/tip/contributing/coding-style-guide.md/#control-structure-prefer-early-returns-and-avoid-else